### ISSUE TITLE
Tensorflow restore checkpoint with monitored session and custom hook failing

### ISSUE THREAD

**Issue by vitalyli** [uber/horovod#348]
Hi @alsrgv Is there any example how to restore from Tensorflow checkpoint? I'm trying to restore it using a custom hook, which triggers on rank0 only, like this one, which errors, see error below:

```
is_chief = hvd.rank() == 0

class MyRestoreCheckpointHook(SessionRunHook):
    def __init__(self, checkpoint):
        super(MyRestoreCheckpointHook, self).__init__()
        self.checkpoint = checkpoint

    def begin(self):
        """Called once before using the session."""
        pass

    def after_create_session(self, session, coord):  # pylint: disable=unused-argument
        #session.run([tf.global_variables_initializer()])
        self.restore_checkpoint(session)

    def before_run(self, run_context):  # pylint: disable=unused-argument
        return None

    def restore_checkpoint(self, session):
        if self.checkpoint is not None:
            print("rank %s restoring from checkpoint: %s" % (self.rank,self.checkpoint))
            ckpt = tf.train.get_checkpoint_state(self.checkpoint)
            if ckpt and ckpt.model_checkpoint_path:
                saver = tf.train.Saver(max_to_keep=10, sharded=True)
                saver.restore(session, ckpt.model_checkpoint_path)
                print("restored from checkpoint: %s" % ckpt.model_checkpoint_path)

hooks = []

if is_chief:
    hooks.append(MyRestoreCheckpointHook("/runs/20180702014654/checkpoint"))
    # Horovod: BroadcastGlobalVariablesHook broadcasts initial variable states
    # from rank 0 to all other processes. This is necessary to ensure consistent
    # initialization of all workers when training is started with random weights
    # or restored from a checkpoint.
    hooks.append(hvd.BroadcastGlobalVariablesHook(0))

hooks.append(tf.train.StopAtStepHook(last_step=200000000 // hvd.size()))
# Horovod: adjust number of steps based on number of GPUs.
hooks.append(tf.train.LoggingTensorHook(tensors={'step': global_step, 'loss': cost}, every_n_iter=100))

with tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir,
                                       hooks=hooks,
                                       config=config) as sess:
 ....
```
INFO:tensorflow:An error was raised while a session was being created. This may be due to a preemption of a connected worker or parameter server. A new session will be created. Error: Horovod has been shut down. This has been caused by an exception on one of the rank or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks has finished execution.
[[Node: HorovodBroadcast_W4_0 = HorovodBroadcastT=DT_FLOAT, root_rank=0, _device="/job:localhost/replica:0/task:0/device:CPU:0"]]
[[Node: HorovodBroadcast_W23_Adam_0/_227 = _Recvclient_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device_incarnation=1, tensor_name="edge_77_HorovodBroadcast_W23_Adam_0", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"]]

Caused by op 'HorovodBroadcast_W4_0', defined at:
File "/gpu_data/runs/20180702173427/script/tensorflow_pttr_tfrec10.py", line 359, in 
train_main(FLAGS)
File "/gpu_data/runs/20180702173427/script/tensorflow_pttr_tfrec10.py", line 252, in train_main
config=config) as sess:
File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 405, in MonitoredTrainingSession
stop_grace_period_secs=stop_grace_period_secs)
File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 816, in init
stop_grace_period_secs=stop_grace_period_secs)
File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 532, in init
h.begin()
File "/usr/local/anaconda3/lib/python3.6/site-packages/horovod/tensorflow/init.py", line 127, in begin
self.bcast_op = broadcast_global_variables(self.root_rank)
File "/usr/local/anaconda3/lib/python3.6/site-packages/horovod/tensorflow/init.py", line 96, in broadcast_global_variables
for var in tf.global_variables()])
File "/usr/local/anaconda3/lib/python3.6/site-packages/horovod/tensorflow/init.py", line 96, in 
for var in tf.global_variables()])
File "/usr/local/anaconda3/lib/python3.6/site-packages/horovod/tensorflow/mpi_ops.py", line 119, in broadcast
return MPI_LIB.horovod_broadcast(tensor, name=name, root_rank=root_rank)
File "", line 94, in horovod_broadcast
File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 328, in apply_op
op_type_name, name, **keywords)
File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
op_def=op_def)
File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3392, in create_op
op_def=op_def)
File "/usr/local/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1718, in init
self._traceback = self._graph._extract_stack() # pylint: disable=protected-access

AbortedError (see above for traceback): Horovod has been shut down. This has been caused by an exception on one of the rank or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks has finished execution.

**Comment by alsrgv** [comment]
@vitalyli, thanks for raising a separate issue.  Are there any errors prior to the `Horovod has been shut down. This has been caused by an exception on one of the rank or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks has finished execution.`?  As the message indicates, if shutdown has been caused by an error, you should see the error in the log before the shutdown message.

**Comment by vitalyli** [comment]
I can re-run it, but that’s the log I see. Basically trying to make restoring from existing checkpoint possible. I had done it before without horovod, so the restore code worked for me in the past. 
There is something that goes against broadcasting of global variables and or the way I have it done from the hook after session creation but before broadcast. Is there some example of restore from checkpoint you could share?

**Comment by alsrgv** [comment]
@vitalyli, can you make a [gist](https://gist.github.com/) the full output beginning from `mpirun` command until the processes exit?  The real error message most likely is hiding in the log.

In general, `MonitoredTrainingSession` will automatically restart from the last checkpoint it can find.  E.g. [this example](https://github.com/uber/horovod/blob/master/examples/tensorflow_mnist.py) does that.

**Comment by vitalyli** [comment]
I will retry. I think the problem is with monitored session. My new run doesn’t have any checkpoint data. It’s the previous run I’m trying to load checkpoint from, which somehow doesn’t work with monitored session and/or the way global variable broadcast works.

**Comment by vitalyli** [comment]
I modified code above to feed new run with checkpoint directory from prior run; removed custom hook and fed checkpoint into MonitoredSession then it started working. Would be good to understand in what order basic session needs to be initialized. Also one more question I had, not matter what I did  global_variables_initializer would lock up and won't work under horovod with NCCL.
I used to be able to call below init_variables_op at session start, currently that is not possible.
    with tf.name_scope("InitVar"):
         init_variables_op = tf.group(tf.global_variables_initializer(),tf.local_variables_initializer(),
name='init_var_op') 


Below code worked however:
   
    checkpoint_dir = "/gpu_data/runs/20180702014654/checkpoint" if hvd.rank() == 0 else None
    hooks = [
        # Horovod: BroadcastGlobalVariablesHook broadcasts initial variable states
        # from rank 0 to all other processes. This is necessary to ensure consistent
        # initialization of all workers when training is started with random weights
        # or restored from a checkpoint.
        hvd.BroadcastGlobalVariablesHook(0),

        tf.train.StopAtStepHook(last_step=200000000 // hvd.size()),
        # Horovod: adjust number of steps based on number of GPUs.
        tf.train.LoggingTensorHook(tensors={'step': global_step, 'loss': cost}, every_n_iter=100)
    ]

    with tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir,
                                           hooks=hooks,
                                           config=config) as sess:

        print("Session start")

        if is_chief:
            train_writer = tf.summary.FileWriter(flags.run_dir + '/train', sess.graph)
            #test_writer = tf.summary.FileWriter(flags.run_dir + '/test', sess.graph)

        #sess.run([init_variables, tf.tables_initializer()])

        #estimate parameters
        trainable_params = np.sum([np.prod(v.shape) for v in tf.trainable_variables()])

        print("Trainable_params", trainable_params)

        while not sess.should_stop():

            try:
                _, g_step = \
                    sess.run([train_op, global_step],
                             feed_dict={x_pkeep: training_dropout,
                                        x_learning_rate: learn_rate})

**Comment by alsrgv** [comment]
@vitalyli, `MonitoredTrainingSession` does session initialization, checkpointing and restore from last checkpoint automatically - so you should not need to manually initialize variables.

**Comment by vitalyli** [comment]
@alsrgv Thanks for the pointers. If you happen to have an example with basic Session without MonitoredTrainingSession magic, I'd like to go back to doing it myself. One other thing I need to figure out is how to run test loop. I used run 10Mil test set through after every 1 epoch of some 200Mil samples and record that using test writer. Whenever I try doing that under horovod, it would report timeout after 60 sec and lead to network locking up. Any pointers on how to accomplish that without breaking all reduce?

**Comment by alsrgv** [comment]
@vitalyli, [this example](https://github.com/uber/horovod/blob/master/examples/tensorflow_word2vec.py) uses pure `tf.Session()`.

As for lock-ups and distributed validation.  Lock-ups warnings are not errors, they just indicate that resources could have been more efficiently utilized, so if it happens every once in a while it's not a big issue.

However, you can (and should, if possible) do distributed validation.  [Here](https://github.com/uber/horovod/blob/master/examples/keras_imagenet_resnet50.py)'s an example with Keras using the metric averaging callback.  You can also implement it yourself with a simple `hvd.allreduce()`.

**Comment by vitalyli** [comment]
Hi @alsrgv Any example of distributed validation without keras ? With ps servers and workers setup, every epoch I’d pause training loop, run 10 mil validation set, record summaries under /valid and continue training loop recording summaries under /train. Somehow with allreduce and monitored training session there are gotchas, that make it more complicated. Do you actually use monitored training session at Uber?

**Comment by alsrgv** [comment]
@vitalyli, not at the moment, but your code in the other thread is on the right track.

**Comment by vitalyli** [comment]
Yes I got it working. I will see if and how I can share it.

**Comment by redBirdTx** [comment]
I had the same problem today. It turns out to a typo on a loss I am trying to monitor....So checking any code change, especially on the monitored properties.

**Comment by czs1886** [comment]
> @vitalyli, `MonitoredTrainingSession` does session initialization, checkpointing and restore from last checkpoint automatically - so you should not need to manually initialize variables.

Hi alsrgv, 
When I restore from a checkpoint file,  it gives me 'Not found: Key global_step not found in checkpoint' Error. Do you have any idea how this happens?

**Comment by czs1886** [comment]
@vitalyli, could you share your way of solving the lock-up problem?
